docs/RFCS/replica_batch.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: replica_batch
+- Status: draft
+- Start Date: 2015-08-11
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+Assuming [#1998](https://github.com/cockroachdb/cockroach/pull/1998) in place, replace `proto.Request` by `proto.BatchRequest` throughout most of the main execution path starting in `Store.ExecuteCmd()`.
+
+# Motivation
+
+[#1998](https://github.com/cockroachdb/cockroach/pull/1998) introduces gateway
+changes after the implementation of which only `BatchRequest` is received by
+`(*Store).ExecuteCmd()`. The changes described here allow `BatchRequest` to
+be submitted to Raft and executed in bulk, which should give significant
+performance improvements, in particular due to the lower amount of Raft-related
+round-trip delays.
+
+# Detailed design
+
+The required changes to the code are plenty.
+
+The sections below follow the main request path and outline the necessary
+changes in each.
+
+## Store.ExecuteCmd
+
+This carries out
+
+* request verification and clock updates, which generalize in a relatively
+  straightforward manner.
+* the store retry loop, and particularly, write intent error handling. This
+  depends on the request which ran into the intent (and whether it's a read
+  or write), so `Replica` needs to return not only an error, but also, for
+  example, the associated index of the request in the `Batch`.
+
+## Replica.AddCmd
+
+Control flow currently splits up into read, write and admin paths. For simplicity,
+allowing `Admin` commands only as single elements of a `Batch`, we can keep the
+admin path intact. Regarding the read/write path, there are two options:
+
+* splitting the `Batch` into sub-batches which are completely read or write only.
+  This has the advantage of possibly less changes in the read and write paths,
+  but requires multiple Raft proposals when reads and writes mix (in the worst
+  case scenario, `len(Batch)-1` of them). Having to bookkeep multiple Raft
+  proposals for a single `Batch` is a disadvantage and raises questions about
+  atomicity and response cache handling.
+* keeping the Batch whole, but merging `(*Replica).add{ReadOnly,Write}Cmd`.
+  The idea is that if we need to go through Raft (i.e. if the `Batch` contains
+  at least one write) anyway, we propose the whole `Batch` and satisfy the
+  reads through `Raft`. If the `Batch` is read-only, it executes directly. It
```

Comment:
I'm actually writing the code. Splitting them up is really, surprisingly difficult. I don't think it's worth it to avoid trouble with response cache.

I agree that two paths for reads is problematic for the reasons you cite: it will be bug prone and bloat the response cache. But I think the right solution is to disallow a mix of reads and writes at the KV client level.